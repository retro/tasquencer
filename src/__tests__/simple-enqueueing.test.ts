import { Effect } from 'effect';
import { it } from 'vitest';

import { Builder, IdGenerator, Service } from '../index.js';
import { getEnabledTaskNames, makeIdGenerator } from './shared.js';

type WorkflowContext = 'complete' | 'fail' | 'cancel';
type WorkItemPayload = WorkflowContext;

const workflowDefinition = Builder.workflow<WorkflowContext>()
  .withName('activities')
  .startCondition('start')
  .task('t1', (t) =>
    t()
      .withWorkItem((w) =>
        w<WorkItemPayload>().onStart(({ startWorkItem, getWorkItem }) =>
          Effect.gen(function* () {
            const { payload } = yield* getWorkItem();
            const {
              enqueueCancelWorkItem,
              enqueueCompleteWorkItem,
              enqueueFailWorkItem,
            } = yield* startWorkItem();
            if (payload === 'cancel') {
              yield* enqueueCancelWorkItem();
            } else if (payload === 'fail') {
              yield* enqueueFailWorkItem();
            } else {
              yield* enqueueCompleteWorkItem();
            }
          })
        )
      )
      .onEnable(({ enableTask }) =>
        Effect.gen(function* () {
          const { enqueueStartTask } = yield* enableTask();
          yield* enqueueStartTask();
        })
      )
      .onStart(({ startTask, getWorkflowContext }) =>
        Effect.gen(function* () {
          const context = yield* getWorkflowContext();
          const { initializeWorkItem, enqueueStartWorkItem } =
            yield* startTask();
          const { id } = yield* initializeWorkItem(context);
          yield* enqueueStartWorkItem(id);
        })
      )
  )
  .endCondition('end')
  .connectCondition('start', (to) => to.task('t1'))
  .connectTask('t1', (to) => to.condition('end'));

it('handles simple enqueueing (1)', ({ expect }) => {
  const program = Effect.gen(function* () {
    const idGenerator = makeIdGenerator();

    const service = yield* workflowDefinition.build().pipe(
      Effect.flatMap((workflow) => Service.initialize(workflow, 'complete')),
      Effect.provideService(IdGenerator, idGenerator)
    );

    yield* service.start();
    const state = yield* service.getState();
    expect(state).toMatchSnapshot();
    expect(getEnabledTaskNames(state)).toEqual(new Set());
    expect(state.workflows[0]?.state).toEqual('completed');
  });

  Effect.runSync(program);
});

it('handles simple enqueueing (2)', ({ expect }) => {
  const program = Effect.gen(function* () {
    const idGenerator = makeIdGenerator();

    const service = yield* workflowDefinition.build().pipe(
      Effect.flatMap((workflow) => Service.initialize(workflow, 'cancel')),
      Effect.provideService(IdGenerator, idGenerator)
    );

    yield* service.start();
    const state = yield* service.getState();
    expect(state).toMatchSnapshot();
    expect(getEnabledTaskNames(state)).toEqual(new Set());
    expect(state.workflows[0]?.state).toEqual('started');
  });

  Effect.runSync(program);
});

it('handles simple enqueueing (3)', ({ expect }) => {
  const program = Effect.gen(function* () {
    const idGenerator = makeIdGenerator();

    const service = yield* workflowDefinition.build().pipe(
      Effect.flatMap((workflow) => Service.initialize(workflow, 'fail')),
      Effect.provideService(IdGenerator, idGenerator)
    );

    yield* service.start();
    const state = yield* service.getState();
    expect(state).toMatchSnapshot();
    expect(getEnabledTaskNames(state)).toEqual(new Set());
    expect(state.workflows[0]?.state).toEqual('failed');
  });

  Effect.runSync(program);
});
