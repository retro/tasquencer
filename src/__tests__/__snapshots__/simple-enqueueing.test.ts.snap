// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handles simple enqueueing (1) 1`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 1,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workItems": [
    {
      "id": "workItem-1",
      "payload": "complete",
      "state": "completed",
      "taskGeneration": 1,
      "taskName": "t1",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workflows": [
    {
      "context": "complete",
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "completed",
    },
  ],
}
`;

exports[`handles simple enqueueing (2) 1`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "started",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workItems": [
    {
      "id": "workItem-1",
      "payload": "cancel",
      "state": "canceled",
      "taskGeneration": 1,
      "taskName": "t1",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workflows": [
    {
      "context": "cancel",
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`handles simple enqueueing (3) 1`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "t1",
      "state": "failed",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workItems": [
    {
      "id": "workItem-1",
      "payload": "fail",
      "state": "failed",
      "taskGeneration": 1,
      "taskName": "t1",
      "workflowId": "workflow-1",
      "workflowName": "activities",
    },
  ],
  "workflows": [
    {
      "context": "fail",
      "id": "workflow-1",
      "name": "activities",
      "parent": null,
      "state": "failed",
    },
  ],
}
`;
